envparse.py

```diff
@@ -1,197 +1,217 @@
-# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
-# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
-# IN THE SOFTWARE.
-
+"""
+envparse is a simple utility to help parsing environment variables.
+"""
+from __future__ import unicode_literals
+from functools import partial
+import inspect
+import json
+import logging
 import os
-import types
-import unittest
+import re
+import shlex
 import warnings
+try:
+    import urllib.parse as urlparse
+except ImportError:
+    # Python 2
+    import urlparse
 
 
-NOTSET = object()
+__version__ = '0.2.0'
 
 
-def env(var, cast=None, default=NOTSET):
-    """Return value for given environment variable.
+logger = logging.getLogger(__file__)
 
-    :param cast: Type to cast return value as.
-    :param default: If var not present in environ, return this instead.
 
-    :returns: Value from environment or default (if set)
-    """
-    try:
-        value = os.environ[var]
-    except KeyError:
-        if default is NOTSET:
-            raise
-
-        value = default
-
-    # Resolve any proxied values
-    if hasattr(value, 'startswith'):
-        if value.startswith('$'):
-            warnings.warn("$foo style proxied values will be removed in 1.0",
-                          DeprecationWarning, stacklevel=2)
-            # Old-style: For backwards compat. Won't work if passed through
-            # the shell without escaping
-            value = value.lstrip('$')
-            value = env(value, cast=cast, default=default)
-        if value.startswith('{') and value.endswith('}'):
-            # New-style: Always works
-            value = value.lstrip('{').rstrip('}')
-            value = env(value, cast=cast, default=default)
-
-    # Don't cast if we're returning a default value
-    if value != default:
-        if cast is bool:
-            value = int(value) != 0
-        elif isinstance(cast, list):
-            value = map(cast[0], [x for x in value.split(',') if x])
-        elif cast:
-            value = cast(value)
-
-    return value
+class ConfigurationError(Exception):
+    pass
 
 
 class Env(object):
-    """Provide schema-based lookups of environment variables so that each
-    caller doesn't have to pass in `cast` and `default` parameters.
+    """
+    Lookup and cast environment variables with optional schema.
 
-    Usage:
+    Usage:::
 
+        env = Env()
+        env('foo')
+        env.bool('bar')
+
+        # Create env with a schema
         env = Env(MAIL_ENABLED=bool, SMTP_LOGIN=(str, 'DEFAULT'))
         if env('MAIL_ENABLED'):
             ...
     """
+    # Can't rely on None since it may be desired as a return value.
+    NOTSET = type(str('NoValue'), (object,), {})
```

Comment:
Minor, but I think I prefer `NOTSET` as a module-level variable rather than a class variable; it shouldn't vary across `Env` implementations. Also, having to do

```python
from envparse import Env

class MyEnv(Env):
    def my_method(self, var, cast=None, default=Env.NOTSET):
        # ...
```

is a bit clumsy.